mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala

```diff
@@ -147,49 +146,40 @@ class RobustScaler (override val uid: String)
 
   override def fit(dataset: Dataset[_]): RobustScalerModel = {
     transformSchema(dataset.schema, logging = true)
-    val localRelativeError = $(relativeError)
 
-    val summaries = dataset.select($(inputCol)).rdd.map {
-      case Row(vec: Vector) => vec
-    }.mapPartitions { iter =>
-      var agg: Array[QuantileSummaries] = null
-      while (iter.hasNext) {
-        val vec = iter.next()
-        if (agg == null) {
-          agg = Array.fill(vec.size)(
-            new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, localRelativeError))
-        }
-        require(vec.size == agg.length,
-          s"Number of dimensions must be ${agg.length} but got ${vec.size}")
-        var i = 0
-        while (i < vec.size) {
-          agg(i) = agg(i).insert(vec(i))
-          i += 1
-        }
-      }
-
-      if (agg == null) {
-        Iterator.empty
-      } else {
-        Iterator.single(agg.map(_.compress))
-      }
-    }.treeReduce { (agg1, agg2) =>
-      require(agg1.length == agg2.length)
-      var i = 0
-      while (i < agg1.length) {
-        agg1(i) = agg1(i).merge(agg2(i))
-        i += 1
-      }
-      agg1
+    val numFeatures = MetadataUtils.getNumFeatures(dataset, $(inputCol))
+    val vectors = dataset.select($(inputCol)).rdd.map {
+      case Row(vec: Vector) =>
+        require(vec.size == numFeatures,
+          s"Number of dimensions must be $numFeatures but got ${vec.size}")
+        vec
     }
 
-    val (range, median) = summaries.map { s =>
-      (s.query($(upper)).get - s.query($(lower)).get,
-        s.query(0.5).get)
-    }.unzip
-
-    copyValues(new RobustScalerModel(uid, Vectors.dense(range).compressed,
-      Vectors.dense(median).compressed).setParent(this))
+    val localRelativeError = $(relativeError)
+    val localUpper = $(upper)
+    val localLower = $(lower)
+
+    val (ranges, medians) = vectors.mapPartitions { iter =>
+      if (iter.hasNext) {
+        iter.flatMap { vec => Iterator.range(0, numFeatures).map(i => (i, vec(i))) } ++
+          // add NaN for each feature as the indication of the end of partition
+          Iterator.range(0, numFeatures).map((_, Double.NaN))
+      } else Iterator.empty
+    }.aggregateByKey(
+      new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, localRelativeError))(
+      // compress the QuantileSummaries at the end of partition
+      seqOp = (s, v) => if (v.isNaN) s.compress else s.insert(v),
+      combOp = (s1, s2) => s1.merge(s2)
+    ).mapValues { s =>
+      val range = s.query(localUpper).get - s.query(localLower).get
+      val median = s.query(0.5).get
+      (range, median)
+    }.sortBy(_._1).values.collect().unzip
```

Comment:
It probably won't matter, but do you want to sort these with Spark, or just sort locally? You collect all the values either way. It might be faster to not do it in Spark.